### PR TITLE
Add translation interface

### PR DESCRIPTION
**Pull request by maciekjablonski**
Refs #815

### CODE REVIEW

File: saleor/core/utils/__init__.py [hunk 1]
```diff
@@ -120,3 +121,23 @@ def create_superuser(credentials):
...
+        for translation in self._instance.translations.all():
+            if translation.language_code == self._local and hasattr(translation, item):
+                return getattr(translation, item)
+
+        return getattr(self._instance, item)
+
+class TranslationProxy(object):
+
+    def __get__(self, instance, owner):
+        local = get_language()
```
Comment (patrys) [inline saleor/core/utils/__init__.py]: Typo?

File: saleor/core/utils/__init__.py [hunk 2]
```diff
@@ -120,3 +121,23 @@ def create_superuser(credentials):
     else:
         msg = 'Superuser already exists - %(email)s' % credentials
     return msg
+
+
+class TranslationWrapper(object):
+
+    def __init__(self, instance, local):
```
Comment (patrys) [inline saleor/core/utils/__init__.py]: Typo?

**Comment by maciekjablonski** [comment]
@patrys review please

File: saleor/product/models/base.py [hunk 3]
```diff
@@ -19,6 +19,7 @@
 from satchless.item import InsufficientStock, Item, ItemRange
 from text_unidecode import unidecode
 
+from saleor.core.utils import TranslationProxy
```
Comment (bogdal) [inline saleor/product/models/base.py]: Please use a relative path.

File: saleor/product/models/base.py [hunk 4]
```diff
@@ -198,6 +200,18 @@ def set_attribute(self, pk, value_pk):
 
 
 @python_2_unicode_compatible
+class ProductTranslation(models.Model):
+    language_code = models.CharField(max_length=5)
+    product = models.ForeignKey(Product, related_name='translations')
+    name = models.CharField(
+        pgettext_lazy('Product field', 'name'), max_length=128)
```
Comment (bogdal) [inline saleor/product/models/base.py]: We don't want to translate other fields?

File: saleor/core/utils/__init__.py [hunk 5]
```diff
@@ -1,6 +1,7 @@
 # coding: utf-8
 from __future__ import unicode_literals
 import decimal
+from django.utils.translation import get_language
```
Comment (bogdal) [inline saleor/core/utils/__init__.py]: Could you sort imports?

File: saleor/product/models/base.py [hunk 6]
```diff
@@ -198,6 +200,22 @@ def set_attribute(self, pk, value_pk):
...
+    description = models.TextField(
+        verbose_name=pgettext_lazy('Product field', 'description'))
+
+    def __str__(self):
+        return '%s - %s' % (self.language_code, self.name)
+
+
+
```
Comment (patrys) [inline saleor/product/models/base.py]: Too many newlines?

File: saleor/product/models/base.py [hunk 7]
```diff
@@ -198,6 +200,22 @@ def set_attribute(self, pk, value_pk):
...
+         verbose_name=pgettext_lazy('Product field', 'language code'))
+    product = models.ForeignKey(Product, related_name='translations',
+        verbose_name = pgettext_lazy('Product field', 'product'))
+    name = models.CharField(
+        pgettext_lazy('Product field', 'name'), max_length=128)
+    description = models.TextField(
+        verbose_name=pgettext_lazy('Product field', 'description'))
+
+    def __str__(self):
+        return '%s - %s' % (self.language_code, self.name)
```
Comment (patrys) [inline saleor/product/models/base.py]: Why is this useful? This objects needs a `__repr__` that reports its PK.

File: templates/product/details.html [hunk 8]
```diff
@@ -66,7 +66,7 @@
             </div>
             <div class="col-md-6 col-12 product__info">
                 <h1 itemprop="name" class="product__info__name">
-                    {{ product }}
+                    {{ product.translated.name }}
```
Comment (patrys) [inline templates/product/details.html]: Can't we pass `product.translated` to templates? I'm not convinced all of the templates need to know about translations.

File: saleor/settings.py [hunk 9]
```diff
@@ -7,6 +7,7 @@
 import dj_email_url
 from django.contrib.messages import constants as messages
 import django_cache_url
+from django.utils.translation import ugettext_lazy as _
```
Comment (bogdal) [inline saleor/settings.py]: Please sort

**Comment by patrys** [comment]
After reviewing this briefly it seems that the product list (the GraphQL API) remains untranslated?

**Comment by maciekjablonski** [comment]
@patrys re-review please

File: saleor/graphql/api.py [hunk 10]
```diff
@@ -28,6 +28,16 @@ def get_ancestors_from_cache(category, context):
         return category.get_ancestors()
 
 
+class TranslatedModel(object):
```
Comment (patrys) [inline saleor/graphql/api.py]: Why is this needed?

File: saleor/core/utils/__init__.py [hunk 11]
```diff
@@ -130,3 +131,28 @@ def create_superuser(credentials):
...
+@python_2_unicode_compatible
+class TranslationWrapper(object):
+
+    def __init__(self, instance, locale):
+        self._instance = instance
+        self._locale = locale
+
+    def __getattr__(self, item):
+        for translation in self._instance.translations.all():
+            if translation.language_code == self._locale and hasattr(translation, item):
```
Comment (patrys) [inline saleor/core/utils/__init__.py]: Shouldn't we do this lookup in `__init__`?

File: saleor/graphql/api.py (hunk 10 above)
Comment (maciekjablonski) [inline saleor/graphql/api.py]: Mixin: use model.translated instead model

Comment (patrys) [inline saleor/graphql/api.py]: But why is it a mixin instead of a subclass of `DjangoObjectType`? And why does it have to check whether a node has a `translated` attribute?

File: saleor/core/utils/__init__.py [hunk 12]
```diff
@@ -130,3 +131,28 @@ def create_superuser(credentials):
...
     return msg
+
+
+@python_2_unicode_compatible
+class TranslationWrapper(object):
+
+    def __init__(self, instance, locale):
+        self._instance = instance
+        self._locale = locale
+        self._translation = instance.translations.filter(language_code=locale).first()
```
Comment (patrys) [inline saleor/core/utils/__init__.py]: This is guaranteed to execute a database query even if translations were properly prefetched.

File: saleor/graphql/api.py (hunk 10 above)
Comment (maciekjablonski) [inline saleor/graphql/api.py]: DjangoObjectType has custom __new__ which requires Meta class

File: saleor/core/utils/__init__.py [hunk 13]
```diff
@@ -130,3 +132,31 @@ def create_superuser(credentials):
...
+@python_2_unicode_compatible
+class TranslationWrapper(object):
+
+    def __init__(self, instance, locale):
+        self._instance = instance
+        self._locale = locale
+
+    @cached_property
+    def _translation(self):
+        return self._instance.translations.filter(language_code=self._locale).first()
```
Comment (patrys) [inline saleor/core/utils/__init__.py]: You can't call filter. Even if you put it behind ten thousand caches. It will execute an extra database query at least once per instance.

**Comment by patrys** [comment]
You're overengineering this. See https://groups.google.com/d/msg/django-developers/6a5Bwf5KMNA/RwRnXc6xBgAJ

**Comment by maciekjablonski** [comment]
@patrys review please

File: templates/product/details.html [hunk 14]
```diff
@@ -68,7 +68,7 @@
     </div>
     <div class="col-md-6 col-12 product__info">
       <h1 class="product__info__name">
-        {{ product }}
+        {{ product.name }}
```
Comment (krzysztofwolski) [inline templates/product/details.html]: Should be translated by str method

File: saleor/product/models/base.py [hunk 15]
```diff
@@ -215,15 +215,19 @@ def get_price_range(self, discounts=None,  **kwargs):
 
 
 class ProductTranslation(models.Model):
-    language_code = models.CharField(max_length=50,
-         verbose_name=pgettext_lazy('Product field', 'language code'))
-    product = models.ForeignKey(Product, related_name='translations',
-        verbose_name = pgettext_lazy('Product field', 'product'))
+    language_code = models.CharField(
+        max_length=50, verbose_name=pgettext_lazy('Product field', 'language code'))
```
Comment (krzysztofwolski) [inline saleor/product/models/base.py]: https://tools.ietf.org/html/rfc5646#section-4.4.1

File: saleor/product/models/base.py [hunk 16]
```diff
@@ -404,7 +404,7 @@ def quantity_available(self):
 class ProductAttribute(models.Model):
     name = models.SlugField(
         pgettext_lazy('Product attribute field', 'internal name'),
-        max_length=50, unique=True)
+        max_length=35, unique=True)
```
Comment (krzysztofwolski) [inline saleor/product/models/base.py]: Why...? Ctrl+R without validation?

**Comment by patrys** [comment]
Let's revisit this.

**Comment by koradon** [comment]
@elwoodxblues pls review

**Comment by Pacu2** [comment]
Work will continue in #2475
